### PR TITLE
Add support for returning callback from `LightningModule.configure_callbacks`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes: see title

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

part of #1

cc @borda @carmocca @justusschock @awaelchli @ananthsub @ninginthecloud @jjenniferdai

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1119,12 +1120,13 @@ def predicts_step(self, batch, batch_idx, dataloader_idx=0):
...
-        present in the Trainer's callbacks list, it will take priority and replace them. In addition, Lightning
-        will make sure :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint` callbacks run last.
+        gets called, the list or a callback returned here will be merged with the list of callbacks passed to the
+        Trainer's ``callbacks`` argument. If a callback returned here has the same type as one or several callbacks
+        already present in the Trainer's callbacks list, it will take priority and replace them. In addition,
+        Lightning will make sure :class:`~pytorch_lightning.callbacks.model_checkpoint.ModelCheckpoint` callbacks
+        run last.
 
         Return:
             A list of callbacks which will extend the list of callbacks in the Trainer.
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1132]: this needs to be updated

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1119,12 +1120,13 @@ def predicts_step(self, batch, batch_idx, dataloader_idx=0):
         """
         return self(batch)
 
-    def configure_callbacks(self):
+    def configure_callbacks(self) -> Optional[Union[Sequence[Callback], Callback]]:
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1123]: why is this optional?

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 3]
```diff
@@ -271,6 +271,9 @@ def _attach_model_callbacks(self) -> None:
         model_callbacks = self.trainer._call_lightning_module_hook("configure_callbacks")
         if not model_callbacks:
             return
+
+        model_callbacks = [model_callbacks] if not isinstance(model_callbacks, Sequence) else model_callbacks
+        model_callbacks = list(model_callbacks)
         model_callback_types = {type(c) for c in model_callbacks}
         trainer_callback_types = {type(c) for c in self.trainer.callbacks}
         override_types = model_callback_types.intersection(trainer_callback_types)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/callback_connector.py:279]: @awaelchli should this use the Callback state_key to disambiguate?

**Review by ananthsub — COMMENTED** [review]
out of curiosity, what motivates this change?

**Comment by rohitgr7** [comment]
> out of curiosity, what motivates this change?

not a change but an enhancement. although nothing to be specific. Just for consistency and support w.r.t `Trainer(callbacks=...)`

**Review by Borda — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -49,6 +49,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added a warning that shows when `max_epochs` in the `Trainer` is not set ([#10700](https://github.com/PyTorchLightning/pytorch-lightning/issues/10700))
 
 
+- Added support for returning a Callback from `LightningModule.configure_callbacks` ([#11060](https://github.com/PyTorchLightning/pytorch-lightning/issues/11060))
```
Comment (awaelchli) [inline CHANGELOG.md:52]:
```suggestion
- Added support for returning a single Callback from `LightningModule.configure_callbacks` without wrapping it into a list ([#11060](https://github.com/PyTorchLightning/pytorch-lightning/issues/11060))
```
Is this maybe clearer?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 5]
```diff
@@ -272,6 +272,9 @@ def _attach_model_callbacks(self) -> None:
         model_callbacks = self.trainer._call_lightning_module_hook("configure_callbacks")
         if not model_callbacks:
             return
+
+        model_callbacks = [model_callbacks] if not isinstance(model_callbacks, Sequence) else model_callbacks
+        model_callbacks = list(model_callbacks)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/callback_connector.py:277]: n00b q: is this list conversion strictly needed? if someone passed a tuple/set of callbacks, is that an issue?

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/callback_connector.py:277]:
```suggestion
```

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/callback_connector.py]: true, not required.
